File: frontend/src/stores/budLinkedFeatures.ts
// Copyright 2025-2026 Arun Rajkumar
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { LinkedFeature, LinkFeaturesResponse } from '@/types'
import api from '@/services/api'
import { extractApiError } from '@/utils/errors'

export const useBudLinkedFeaturesStore = defineStore('budLinkedFeatures', () => {
  const byBudId = ref<Record<string, LinkedFeature[]>>({})
  const loading = ref(false)
  const error = ref('')

  async function fetch(budId: string): Promise<LinkedFeature[] | null> {
    loading.value = true
    error.value = ''
    try {
      const { data } = await api.get<LinkedFeature[]>(`/v1/buds/${budId}/linked-features`)
      byBudId.value = { ...byBudId.value, [budId]: data }
      return data
    } catch (err) {
      error.value = extractApiError(err, 'Failed to load linked features')
      return null
    } finally {
      loading.value = false
    }
  }

  async function link(
    budId: string,
    featureIds: string[],
  ): Promise<LinkFeaturesResponse | null> {
    error.value = ''
    try {
      const { data } = await api.post<LinkFeaturesResponse>(
        `/v1/buds/${budId}/linked-features`,
        { featureIds },
      )
      await fetch(budId)
      return data
    } catch (err) {
      error.value = extractApiError(err, 'Failed to link features')
      return null
    }
  }

  async function unlink(budId: string, featureId: string): Promise<boolean> {
    error.value = ''
    try {
      await api.delete(`/v1/buds/${budId}/linked-features/${featureId}`)
      await fetch(budId)
      return true
    } catch (err) {
      error.value = extractApiError(err, 'Failed to unlink feature')
      return false
    }
  }

  return { byBudId, loading, error, fetch, link, unlink }
})

File: backend/app/schemas/bud_linked_features.py
# Copyright 2025-2026 Arun Rajkumar
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Pydantic schemas for the BUD↔feature linkage endpoints.

Split out of :mod:`app.schemas.bud` so the linked-feature DTOs live next
to their endpoint (``api/v1/bud_linked_features.py``) rather than in the
core BUD CRUD module. The shape is camelCase via ``model_config`` so the
frontend consumes it without renaming.
"""

import uuid

from pydantic import BaseModel, Field


class LinkedFeatureRead(BaseModel):
    """One feature linked to a BUD, with PRIMARY-repo metadata flattened.

    Shape is camelCase via ``model_config`` aliases so the frontend can
    consume it without renaming. ``code_locations`` is the JSONB blob
    from the PRIMARY :class:`FeatureToRepo` row — null when no PRIMARY
    junction exists (rare; happens for legacy or BUD-authored features).
    """

    id: uuid.UUID
    title: str = Field(..., alias="title")
    link_type: str = Field(..., alias="linkType")
    source: str
    repo_id: uuid.UUID | None = Field(default=None, alias="repoId")
    repo_name: str | None = Field(default=None, alias="repoName")
    code_locations: dict[str, list[str]] | None = Field(default=None, alias="codeLocations")

    model_config = {"populate_by_name": True}


class LinkFeaturesRequest(BaseModel):
    """Body for POST /buds/{bud_id}/linked-features.

    ``feature_ids`` is treated as a set — duplicates are dropped, and
    ids that don't belong to the requesting org or are inactive are
    silently filtered at the repository layer.
    """

    feature_ids: list[uuid.UUID] = Field(..., min_length=1, alias="featureIds")

    model_config = {"populate_by_name": True}


class LinkFeaturesResponse(BaseModel):
    """Response for POST /buds/{bud_id}/linked-features."""

    inserted_count: int = Field(..., alias="insertedCount")
    inserted_feature_ids: list[uuid.UUID] = Field(..., alias="insertedFeatureIds")

    model_config = {"populate_by_name": True}
